src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const { type, keysToRemove, extraData } of actions) {
+    switch (type) {
+      case 'addProperties':
+        addProps(currentState, extraData);
+        break;
+      case 'removeProperties':
+        removeProps(currentState, keysToRemove);
+        break;
+      default:
+        currentState = {};
+    }
+
+    stateHistory.push({ ...currentState });
```

Comment:
It's unnecessary to spread `currentState` into a new object here since you should be creating a new state object inside the `addProps` and `removeProps` functions. If those functions are corrected to return new objects, you can directly push the returned state into `stateHistory`.